File: project/pmsm_simulation/cpp/src/pmsm.cc
#include "pmsm.h"

#include <eigen3/Eigen/Dense>

using namespace Eigen;

/* Electric macnine function definitions */
MatrixXd PMSM::fx(const MatrixXd &x, const MatrixXd &u) { // State equations
  // extract states
  double id = x(0);
  double iq = x(1);
  double omega = x(2);

  // extract inputs
  double vd = u(0);
  double vq = u(1);
  double Tl = u(2);

  // state equations
  double id_dot =
      1 / Ld * (vd - Rs * id + np * omega * Lq * iq); // d-axis current update
  double iq_dot =
      1 / Lq *
      (vq - Rs * iq - np * omega * (Ld * id + psi_r)); // q-axis current update
  double omega_dot = 1 / J *
                     (3 * np / 2 * (psi_r * iq + (Ld - Lq) * id * iq) -
                      Tl); // mechanical speed update

  xdot << id_dot, iq_dot, omega_dot;

  return xdot;
}

void PMSM::set_time(const double &start, const double &end,
                    const double &delta) { // set time vector
  t = VectorXd::LinSpaced((end - start) / delta, start, end);
}

void PMSM::simulate(const VectorXd &x0) { // simulate the PMSM<
  x = MatrixXd::Zero(3, t.size());
  x.col(0) = x0;

  RungeKutta4<PMSM>(t, x, u, *this);
}

void PMSM::set_inputs(const int &input_idx, const double &val,
                      const double &t0) { // set input vector
  u = MatrixXd::Zero(3, t.size());
  u.row(input_idx) = UnitStep(t0) * val;
}

VectorXd PMSM::UnitStep(const double &t0) { // unit step function
  VectorXd u = VectorXd::Zero(t.size());    // initialize with zeros
  for (int ii = 0; ii < t.size(); ++ii) {
    if (t[ii] > t0) {
      u[ii] = 1;
    }
  }
  return u;
}

File: project/pmsm_simulation/matlab/PMSM.m
%% creating a PMSM class
classdef PMSM
    properties
        % motor parameters
        Ld = 2.85e-3; % d-axis inductance
        Lq = 2.85e-3; % q-axis inductance
        Rs = 0.054; % stator resistance 
        psi_r = 0.8603; % rotor flux linkage
        np = 3; % number of pole pairs
        J = 0.25; % rotor inertia
        
        % t; % time vector
        t = 0; % time vector
        u = 0; % input vector
        x = 0; % state vector
        integrator = 0; % solver object
        % u; % input vector
        % x; % state vector
        % integrator; % solver object
    end
    methods
        function obj = PMSM(Ld, Lq, Rs, psi_r, np, J)
            % constructor
            if nargin > 0
                obj.Ld = Ld;
                obj.Lq = Lq;
                obj.Rs = Rs;
                obj.psi_r = psi_r;
                obj.np = np;
                obj.J = J;
            end
        end

        function obj = set_time(obj, start, finish, step)
            obj.t = start:step:finish;
        end

        function obj = set_input(obj, idx, val, t0)
            obj.u = zeros(3, length(obj.t));
            obj.u(idx, obj.t >= t0) = val;
        end
        
        function xdot = fx(obj, x, u)
            % state-space model of PMSM
            % x = [id; iq; omega]
            % u = [v_d; v_q]
            id = x(1);
            iq = x(2);
            omega = x(3);
            vd = u(1);
            vq = u(2);
            Tl = u(3);
            
            % d-axis voltage equation
            id_dot = (1 / obj.Ld) * (vd - obj.Rs * id + obj.np * omega * obj.Lq * iq);
            iq_dot = (1 / obj.Lq) * (vq - obj.Rs * iq - obj.np * omega * (obj.Ld * id + obj.psi_r));
            omega_dot = (1 / obj.J) * ( 3 * obj.np / 2 * (obj.psi_r * iq + (obj.Ld - obj.Lq) * id * iq) - Tl); 

            xdot = [id_dot; iq_dot; omega_dot];
        end

        function obj = simulate(obj, x0)
            % initial conditions
            obj.x = zeros(3, length(obj.t));
            obj.x(:, 1) = x0;

            obj.integrator = RunkeKutta4();
            obj.integrator = obj.integrator.set_func(@(x, u) obj.fx(x, u));
            obj.x = obj.integrator.integrare(obj.t, obj.x, obj.u);
        end 
    end
end